xarray/core/arithmetic.py

```diff
@@ -35,7 +35,9 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
         # See the docstring example for numpy.lib.mixins.NDArrayOperatorsMixin.
         out = kwargs.get("out", ())
         for x in inputs + out:
-            if not isinstance(x, self._HANDLED_TYPES + (SupportsArithmetic,)):
+            if not isinstance(
+                x, self._HANDLED_TYPES + (SupportsArithmetic,)
+            ) and not hasattr(x, "__array_function__"):
```

Comment:
should this be `__array_ufunc__` instead?

Reply:
I'm no expert but to my reading of [NEP18](https://numpy.org/neps/nep-0018-array-function-protocol.html#backward-compatibility), yes. Then once `DataArray.array_function` is defined it can check the instance for presence of `__array_ufunc__` and `__array_function__` attributes separately.

Reply:
Thanks, @amcnicho. `DataArray`, `Dataset` and `Variable` inherit from `SupportsArithmetic`, so they are included anyways (no support for `__array_function__` on `xarray` objects, yet, that would be a different function, see also #3917). Rereading the code I'm pretty sure you're right; we need `__array_ufunc__` instead. I'll change it.